### PR TITLE
added qft

### PR DESCRIPTION
**Pull request by eurethia**
I don't know if this would work. Here's an image from jupyter notebook before I turn my code into the function. 
image

### CODE REVIEW

**Comment by refraction-ray** [comment]
seems to me correct though there are some unused circuit lines such as  0,1,4? For a more detailed intro on QFT, you can refer my [post](https://re-ra.xyz/%E7%BB%8F%E5%85%B8%E4%B8%8E%E9%87%8F%E5%AD%90%E7%9A%84%E5%BF%AB%E9%80%9F%E5%82%85%E7%AB%8B%E5%8F%B6%E5%8F%98%E6%8D%A2/)

**Comment by eurethia** [comment]
If the user enter [0,1,4] as index, then these circuit lines will be used.
image

**Comment by refraction-ray** [comment]
mypy fails now in GitHub action, function is missing type annotation

**Review by refraction-ray — CHANGES_REQUESTED** [review]
See comments

File: tensorcircuit/templates/blocks.py [hunk 1]
```diff
@@ -149,3 +149,19 @@ def example_block(
...
+    for q in index:
+        if index.count(q) > 1:
+            raise TypeError("There should not be any repetitive qubits")
+    for i in range(len(index)):
+        c.H(index[i])
+        rotation = np.pi / 2
+        for j in range(i + 1, len(index)):
+            c.cphase(index[j], index[i], theta=rotation)
+            rotation /= 2
+        c.barrier_instruction()
```
[the remark this answers is not in the document]
Comment (eurethia) [reply tensorcircuit/templates/blocks.py:164]: okay since it's useless I will delete this

File: tensorcircuit/templates/blocks.py [hunk 2]
```diff
@@ -149,3 +149,25 @@ def example_block(
...
+
+
+def qft(c: Circuit, *index: int) -> None:
+    """
+    This function applies quantum fourier transformation (QFT) to the selected circuit lines
+
+    :param c: Circuit in
+    :type c: Circuit
+    :param *index: the indices of the circuit lines to apply QFT
+    :type *index: Tuple[int, int]
```
Comment (refraction-ray) [inline tensorcircuit/templates/blocks.py:161]: the type of index can be documented as int or List[int], but not Tuple[int, int], as the latter only refer two cases with two ints

Comment (refraction-ray) [reply tensorcircuit/templates/blocks.py:161]: also better return the circuit c instead of None for consistency with other template function

**Comment by refraction-ray** [comment]
not github action fails black, please run `check_all.sh` locally at least up to pylint stage to ensure the format of the code

**Comment by refraction-ray** [comment]
Also I strongly recommend to add the following options to the function, these can be done is this PR or next, either is fine for me.

```
do_swaps (bool) – Whether to include the final swaps in the QFT.

inverse (bool) – If True, the inverse Fourier transform is constructed.

insert_barriers (bool) – If True, barriers are inserted as visualization improvement.
```

File: tensorcircuit/templates/blocks.py (hunk 2 above)
Comment (eurethia) [reply tensorcircuit/templates/blocks.py:161]: ok

**Comment by eurethia** [comment]
The black test passed on my PC.

File: tensorcircuit/templates/blocks.py [hunk 3]
```diff
@@ -149,3 +149,54 @@ def example_block(
         for i in range(n):
             c.rx(i, theta=param[2 * j + 1, i])
     return c
+
+
+def qft(
+    c: Circuit,
+    *index: int,
+    do_swaps: bool = False,
```
Comment (refraction-ray) [inline tensorcircuit/templates/blocks.py:157]: this should default to True, as it is part of the QFT

Comment (eurethia) [reply tensorcircuit/templates/blocks.py:157]: Sure

**Review by refraction-ray — APPROVED** [review]
LGTM

**Comment by refraction-ray** [comment]
thanks for the nice contribution! @all-contributors please add @eurethia for code, test

**Pull request merged**